Modifies LcmPublisherSystem to optionally throttle its publish rate.

liangfok opened the pull request:
Resolves #4502.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4731)
<!-- Reviewable:end -->

jwnimmer-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-Ka8vnpppYwID_rqih-Q:-Ka8vnpqxSVONcxxRk1a:bd59nmu) (no related file):*
Further overloading the constructors / factory methods seems like a code smell.  Why not just have a `set` method to change the publication period, separate from construction?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-Ka8vnpppYwID_rqih-Q:-Ka8ycvWcSODrL_lm395:b-pav9e1) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Further overloading the constructors / factory methods seems like a code smell.  Why not just have a `set` method to change the publication period, separate from construction?
</blockquote></details>

SGTM. Will update this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@jwnimmer-tri for feature review please.
+@sammy-tri for platform review please.

Thanks!

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaAmQiftcFxnG0RM5N7:bjege30)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAoEJPOFSkD-XoOwyd:-KaAoEJPOFSkD-XoOwye:bb1em88) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L164)):*
> ```objc
> 
>   // Last publish time.
>   mutable double last_publish_time_{};
> ```

Dear reviewers, 

I'm particularly worried about the need to make this `mutable`. Should this be stored in the `Context`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@liangfok, since this is triggering off the simulated time why not use a periodic publishing rate via `LeafSystem::DeclarePublishPeriodSec()`rather than trying to slow down an every-timestep rate?

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaAuLsRhK67rpN3Iha2:bfs87dw)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 5 files at r1.
Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAvweDCu2OkY6MndaR:-KaAvweDCu2OkY6MndaS:bxcf0g6) (no related file):*
Right.  So DeclarePublishPeriodSec seems be exactly what is wanted, without anything useful left in this PR?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAsEWYcAa3c6DxEEk0:-KaAsEWYcAa3c6DxEEk1:b6btsba) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L26)):*
> ```objc
>  public:
>   /**
>    * A factory method that returns an LcmPublisherSystem that takes
> ```

*Meta* for all of these ctors -- wasn't the doxygen auto-link the System base class more useful that a self-link?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaArj3Lb2HqKJbHB57O:-KaArj3McUYbcAzkrTgR:b-ahgk7r) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L29)):*
> ```objc
>    * Value<LcmMessage> message objects on its sole abstract-valued input port.
>    * The returned LcmPublisherSystem publishes every major timestep of the
>    * simulation.
> ```

I think this sentence added to all ctors docs is perhaps an oversight from when the ctor used to be overloaded?  Or in any case, I think brittle duplicating a comment about the default is probably defective; the class overview note should be enough I think?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAtDKY2-7ZsOxpBhSq:-KaAtDKY2-7ZsOxpBhSr:bnfv3ah) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L109)):*
> ```objc
>   const std::string& get_channel_name() const;
> 
>   // Sets the publishing period of this system. The publishing period is the
> ```

*Meta* I think this API comment, and the one below, should use Doxgen comment blocks (e.g., `///`).

---

*[drake/systems/lcm/lcm_publisher_system.h, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAtj8mLcrcdtXypmiU:-KaAtj8mLcrcdtXypmiV:b-e38g8e) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L112)):*
> ```objc
>   // amount of simulation time that elapses between successive publish
>   // operations. A value of infinity results in this system publishing at every
>   // major timestep of the simulation.
> ```

BTW I probably would have gone with `0` and not `+inf` as the magic value, but I guess neither one is great anyway, so might as well just stick with `+inf`.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAu8CEyzRJve-QSBOa:-KaAu8CFk-pkfwUAL6jH:bkvwt7u) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L154)):*
> ```objc
> 
>   void DoCalcNextUpdateTime(const Context<double>& context,
>                             UpdateActions<double>* actions) const override;
> ```

BTW I think this method is `protected:` in the base class that this is overriding.  We typically preserve that access in derived classes (rather than narrowing it), but I'm not sure it's a rule per se.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAoEJPOFSkD-XoOwyd:-KaAvjmQPRwBTOlTqHeT:bi0n62p) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L164)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Dear reviewers, 

I'm particularly worried about the need to make this `mutable`. Should this be stored in the `Context`?
</blockquote></details>

Wait, this reimplements event logic from System and LeafSystem?  I guess I'll post a whole-PR discussion on that and stop reviewing now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAvweDCu2OkY6MndaR:-KaAwFGJnTTnrp8tI3YV:bh94kha) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Right.  So DeclarePublishPeriodSec seems be exactly what is wanted, without anything useful left in this PR?
</blockquote></details>

I guess the `lcm_publisher_system_test.cc` tidying should be retained.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
In #4502's description, @david-german-tri typed:

> This is now possible in System2/Simulator2 using CalcNextUpdateTime.

Is the reference to `CalcNextUpdateTime()` old, i.e., perhaps prior to the existence of `LeafSystem::DeclarePublishPeriodSec()`?

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaCQzpC-KVE41UxVs5Z:b2k4ib8)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
DeclarePublishPeriodSec is syntactic sugar that configures CalcNextUpdateTime for fixed-period updates.  It appears to suffice for what you're doing here!

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAoEJPOFSkD-XoOwyd:-KaD0krCBCREIeFiPlvG:b-3br3lk) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Wait, this reimplements event logic from System and LeafSystem?  I guess I'll post a whole-PR discussion on that and stop reviewing now.
</blockquote></details>

BTW, yes, you should just use DeclareUpdatePeriodSec.  However, modulus arithmetic would avoid the need for this `mutable` member (as is done in LeafSystem itself).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaD0Ez_6YVnS1uyRXq5:b-wtacry)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAvweDCu2OkY6MndaR:-KaCn9b6UHU8J_FR6MVq:bbwwmvp) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess the `lcm_publisher_system_test.cc` tidying should be retained.
</blockquote></details>

I extracted out the tidying of `lcm_publisher_system_test.cc` into #4743.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAsEWYcAa3c6DxEEk0:-KaCltclgScecg_4KZ56:bymutb4) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

*Meta* for all of these ctors -- wasn't the doxygen auto-link the System base class more useful that a self-link?
</blockquote></details>

I modified the Doxygen comment to not include a self link.  I think "LcmPublisherSystem" is more precise and thus less confusing than "System" in this context.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaArj3Lb2HqKJbHB57O:-KaCsE6zTr9-V1pv1Qw_:b-fzj022) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think this sentence added to all ctors docs is perhaps an oversight from when the ctor used to be overloaded?  Or in any case, I think brittle duplicating a comment about the default is probably defective; the class overview note should be enough I think?
</blockquote></details>

Done (removed).

---

*[drake/systems/lcm/lcm_publisher_system.h, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAtDKY2-7ZsOxpBhSq:-KaCsGXvEnVziFyiGXaC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L109)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

*Meta* I think this API comment, and the one below, should use Doxgen comment blocks (e.g., `///`).
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 112 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAtj8mLcrcdtXypmiU:-KaCsI8dWj6FCK2sarj0:bm11j74) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I probably would have gone with `0` and not `+inf` as the magic value, but I guess neither one is great anyway, so might as well just stick with `+inf`.
</blockquote></details>

Agreed. I filed #4747 so we can formally decide whether to make the change.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAu8CEyzRJve-QSBOa:-KaCtX5uDyeF3RfWTrlT:b-g2s2hw) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L154)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I think this method is `protected:` in the base class that this is overriding.  We typically preserve that access in derived classes (rather than narrowing it), but I'm not sure it's a rule per se.
</blockquote></details>

I removed this override.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 164 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaAoEJPOFSkD-XoOwyd:-KaCusP4XcCq3MyiGnlX:b-tq0hd0) ([raw file](https://github.com/robotlocomotion/drake/blob/0c2bddb192f4070dce81d861e5b7b6ef84d13ed6/drake/systems/lcm/lcm_publisher_system.h#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Wait, this reimplements event logic from System and LeafSystem?  I guess I'll post a whole-PR discussion on that and stop reviewing now.
</blockquote></details>

I updated this PR to use `LeafSystem<double>::DeclarePublishPeriodSec(period)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaCnEiUpXyD1XzeXZD4:bblnx7i)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint.

---

Reviewed 2 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDIfvsZPFHFJuadlcF:-KaDIfvsZPFHFJuadlcG:bmoack3) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L20)):*
> ```objc
>  * Publishes an LCM message containing information from its input port. The
>  * default behavior of this system is to publish at every major simulation
>  * timestep. To tune the publishing period, use set_publish_period().
> ```

The default "publish every timestep" is a bit misleading.  The Simulator::set_publish_every_timestmp is what owns that decision, not this class.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 45 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDGa2hx9D-ZV_Oaakr:-KaDGa2iXxcx6xeUHRDo:b-9w58fj) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L45)):*
> ```objc
>    * A constructor for an %LcmPublisherSystem that takes LCM message objects on
>    * its sole abstract-valued input port and publishes at every major timestep
>    * of the simulation. The LCM message type is determined by the provided
> ```

*Meta* This (and several below) still say "and publishes at every major timestep of the simulation" which is both wrong, and brittlely repetitive.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHMpjjqubzvU-4pgx:-KaDHMpkvthGyK2RfzoR:b-g2123e) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L89)):*
> ```objc
>    * @param[in] translator_dictionary A dictionary for obtaining the appropriate
>    * translator for a particular LCM channel. This reference must remain
>    * valid for the lifetime of this object.
> ```

I don't think this new text is true?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHUFsUchj784KO2e_:-KaDHUFvppWNq2kBo00d:b-5xzg27) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L110)):*
> ```objc
>   /// operations. A value of infinity results in this system publishing at every
>   /// major timestep of the simulation.
>   void set_publish_period(double period);
> ```

BTW What do you think about keeping the method named the same thing as the on the LeafSystem?  I think you can even just `using` it in that case, without having to write a wrapper.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 112 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHxWtLsHADRaHDQ_L:-KaDHxWuZwfWvMvk8YjU:b-b5y02j) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L112)):*
> ```objc
>   void set_publish_period(double period);
> 
>   // TODO(liang.fok) Restore this method once #4746 is addressed.
> ```

Commented-code is gross.  You can implement this method now if you want (just stash the value set_publish_period is called with).  Or removing it is fine also -- I'm not sure there's really any need for it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaDJ1xt6dfZchTbVYSO:by96gdd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/lcm/lcm_publisher_system.h, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHUFsUchj784KO2e_:-KaDKW75kxvH1m8Rqldu:b-3wv479) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW What do you think about keeping the method named the same thing as the on the LeafSystem?  I think you can even just `using` it in that case, without having to write a wrapper.
</blockquote></details>

FYI David says he likes this approach as-written better, so _shrug_.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHMpjjqubzvU-4pgx:-KaDfOiM4jZ4C4SvhcIJ:btvshw0) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't think this new text is true?
</blockquote></details>

I guess it's one way of ensuring that the underlying translator remains valid, and likely the simplest.  So defect retracted.

But now the Publisher and Subscriber docs differ in this regard; so fix the Subscriber docs as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDg8SJ1y1BJf7lSGOK:-KaDg8SKhnMklkK9tEkB:b-6ghdd) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L1)):*
> ```cpp
> #include "drake/systems/lcm/lcm_publisher_system.h"
> ```

So I haven't read this file yet, but given that the method is only one line long, it seems like an unreasonable amount of test code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Looks like I got to this one after substantial revision.  I don't think I have anything meaningful to add beyond the issues which are still open.  I'll <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> to get out of the way so that you can merge once those are fixed.

---

Reviewed 1 of 5 files at r1, 2 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaDrn7OPL1SACMzL5_1:b-6nl6xa)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDIfvsZPFHFJuadlcF:-KaEMUhCu36fjYdY2vFX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L20)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The default "publish every timestep" is a bit misleading.  The Simulator::set_publish_every_timestmp is what owns that decision, not this class.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 45 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDGa2hx9D-ZV_Oaakr:-KaEMWQb7niVqfM7hYeM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L45)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

*Meta* This (and several below) still say "and publishes at every major timestep of the simulation" which is both wrong, and brittlely repetitive.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 89 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHMpjjqubzvU-4pgx:-KaEMbn0mEmHvzHeX3oU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess it's one way of ensuring that the underlying translator remains valid, and likely the simplest.  So defect retracted.

But now the Publisher and Subscriber docs differ in this regard; so fix the Subscriber docs as well?
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 110 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHUFsUchj784KO2e_:-KaEMddgwR8sL_rkK6p8:bb1r392) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L110)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI David says he likes this approach as-written better, so _shrug_.
</blockquote></details>

OK. Will keep as is.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 112 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDHxWtLsHADRaHDQ_L:-KaEMfcjPncDfsbuGfIP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/lcm_publisher_system.h#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Commented-code is gross.  You can implement this method now if you want (just stash the value set_publish_period is called with).  Or removing it is fine also -- I'm not sure there's really any need for it.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDg8SJ1y1BJf7lSGOK:-KaEMwI7j2Nlye0oylkN:b-wlqbdq) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

So I haven't read this file yet, but given that the method is only one line long, it seems like an unreasonable amount of test code.
</blockquote></details>

Please take a look. Let me know if I should remove it from this PR. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF3HOzMQUe3eMFxJfs:-KaF3HOzMQUe3eMFxJft:bsdiilm) (no related file):*
This PR has gone through so many rewrites that I'm sure that I'm blind to some defects by now.  Once the current discussions are resolved, please invite David or Sherm as an extra, fresh reviewer.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF0J8hhx2aS1Z4J4st:-KaF0J8hhx2aS1Z4J4su:b2qm5vl) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L86)):*
> ```objc
>    * valid for the lifetime of this object.
>    *
>    * @param lcm A non-null pointer to the LCM subsystem to publish on. This
> ```

BTW The "This pointer" vs "The pointer" is now inconsistent in the different constructors and factory.  And FWIW, I think "The" was better, because "This" means `this` (the method receiver) sometimes.  But at least be consistent within the file one way or the other if you are going to rephrase it.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 104 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF1iZECumHx2NLN3ea:-KaF1iZF5f_KETaRCYce:bs3nu1d) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L104)):*
> ```objc
>   /// amount of simulation time that elapses between successive publish
>   /// operations. A value of infinity results in this system publishing at every
>   /// major timestep of the simulation.
> ```

The final sentence is not quite true, I think.  Infinity means "whatever the simulator does by default" or something, not "every major timestep". 

Perhaps consider, rather than rewriting some API language and getting it wrong and/or being brittle, maybe just specify this method by citing the LeafSystem method's contract that it exposes?  It's not like we're transforming or hiding the semantics; we are merely electing to allow callers to govern this, rather than keeping it encapsulated within the System, as e.g., a device driver would want to do.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF2-XY_w-28IgccM3K:-KaF2-XY_w-28IgccM3L:b-s46y7q) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L108)):*
> ```objc
> 
>   // TODO(liang.fok) Update this to use a LeafSystem accessor method once #4746
>   // is addressed.
> ```

BTW Shouldn't the TODO be placed at the implementation, not the declaration?

---

*[drake/systems/lcm/lcm_publisher_system.h, line 113 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF275f7M0sEqCgmKxc:-KaF275f7M0sEqCgmKxd:bq1lav9) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L113)):*
> ```objc
>   /// operations. A value of infinity means this system publishes at every major
>   /// timestep of the simulation.
>   double get_publish_period() const;
> ```

I'm not seeing where this method is defined?  Which I guess means its not unit tested either?

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaDg8SJ1y1BJf7lSGOK:-KaF4-cxsVSGCiSX8YhQ:b-sdfr90) ([raw file](https://github.com/robotlocomotion/drake/blob/740a99bf8bcdfe50d1fbed03a4cf5e197f67d92b/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L1)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Please take a look. Let me know if I should remove it from this PR. Thanks!
</blockquote></details>

Well the idea wasn't to remove it, just to trim it down to useful content.  I'll add specific notes below and resolve this particular discussion.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 69 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4CNPCSmBvuG0pERw:-KaF4CNPCSmBvuG0pERx:bkk9g7j) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L69)):*
> ```cpp
>   DRAKE_ASSERT(next_t > kSimTime);
>   EXPECT_EQ(update_actions.time, next_t);
> }
> ```

This case is a lot of code just to re-confirm that the LeafSystem code is correct.  It doesn't seem worth it to me.  Convince me otherwise?

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 104 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4eSpSPgvejmRjEDh:-KaF4eSpSPgvejmRjEDi:b2vntzv) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L104)):*
> ```cpp
>   for (int i = 0; i < kDim; ++i) {
>     vector_value[i] = i;
>   }
> ```

BTW Not a big deal, but other tests already test the correctness of the message content; we just want to test the timestamp here, and the particular content here is never validated in this case anyway.

It would be slightly briefer to just use some `Eigen::Zero`, e.g., `context->FixInputPort(kPortNumber, make_unique<BasicVector<double>>(Eigen::Zero(kDim)));`.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4Ph4vIjZKeQ2jPQo:-KaF4Ph4vIjZKeQ2jPQp:b23mv4k) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L112)):*
> ```cpp
>   simulator.set_publish_every_time_step(false);
>   simulator.reset_integrator<systems::RungeKutta3Integrator<double>>(*dut,
>                                               simulator.get_mutable_context());
> ```

Unclear to me why a non-default integrator is required for this unit test.  There is no state?

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 115 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF6s2WDWVdkiMHVhzj:-KaF6s2XkYMnnueuCOEI:btk267u) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L115)):*
> ```cpp
>   simulator.get_mutable_integrator()->set_fixed_step_mode(true);
>   simulator.get_mutable_integrator()->set_minimum_step_size(1e-3);
>   simulator.get_mutable_integrator()->set_maximum_step_size(1e-3);
> ```

Unclear to me why the integration parameters should be relevant at all?  The publish times better be correct no matter what any integrator is doing.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 156 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF6QcQkSB-2WkZkW2Y:-KaF6QcRCTRBo7nXS6Ag:b-57eodv) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L156)):*
> ```cpp
>   simulator.StepTo(3.0);
>   EXPECT_EQ(simulator.get_mutable_context()->get_time(), 3.0);
>   VerifyTimestamp(lcm.get_last_published_message(channel_name), 3.0 * 1000);
> ```

These three lines are repeated 8 times and are all very copy-pasta'y.  Convince me that this phrasing is better than a loop over a data table, or even just checking that `(timestamp * 1e-3) == floor(get_time / kPublishPeriod)` during a for-loop-swept range of StepTo targets?  That would be like 5 lines of code.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF7hvdeFz3W8n3J4RW:-KaF7hvdeFz3W8n3J4RX:b-jlcltf) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L162)):*
> ```cpp
> }  // namespace lcm
> }  // namespace systems
> }  // namespace drake
> ```

Once this is down to a tiny unit test (or really even the two that are here now), it's unclear to me why these unit tests live in a separate file.  That's atypical and will hinder discoverability, with no upside that I can determin.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/lcm/lcm_publisher_system.h, line 113 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF275f7M0sEqCgmKxc:-KaF89XZI4S7O4koP_1k:b-aq40a8) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I'm not seeing where this method is defined?  Which I guess means its not unit tested either?
</blockquote></details>

Again, probably just worth deleting this, since nobody needs it!  They can write and test it if and when they do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Partial address of reviewer comments.

---

Review status: 4 of 5 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 69 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4CNPCSmBvuG0pERw:-KaFHN5Q3HFpmvOoFPib:bl584ku) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This case is a lot of code just to re-confirm that the LeafSystem code is correct.  It doesn't seem worth it to me.  Convince me otherwise?
</blockquote></details>

It's definitely true that `LeafSystem` must be correct for this unit test to pass.  However, isn't there some benefit to protect `LcmPublisherSystem` against future changes in `LeafSystem`, which this unit test perhaps does?

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 104 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4eSpSPgvejmRjEDh:-KaFIB8l9noEsR4iO_fX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L104)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Not a big deal, but other tests already test the correctness of the message content; we just want to test the timestamp here, and the particular content here is never validated in this case anyway.

It would be slightly briefer to just use some `Eigen::Zero`, e.g., `context->FixInputPort(kPortNumber, make_unique<BasicVector<double>>(Eigen::Zero(kDim)));`.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 112 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4Ph4vIjZKeQ2jPQo:-KaFIG19b3-2BfxkIq-s:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L112)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear to me why a non-default integrator is required for this unit test.  There is no state?
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 115 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF6s2WDWVdkiMHVhzj:-KaFIJ29AnGWN_lD4fhx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L115)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear to me why the integration parameters should be relevant at all?  The publish times better be correct no matter what any integrator is doing.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF7hvdeFz3W8n3J4RW:-KaFMPB_i4MGRp-ckPJB:b-ouhzbm) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L162)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Once this is down to a tiny unit test (or really even the two that are here now), it's unclear to me why these unit tests live in a separate file.  That's atypical and will hinder discoverability, with no upside that I can determin.
</blockquote></details>

I'll move these unit tests into the other file once they are done being reviewed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaFMosnxCUcqUafEaKi:b5vhxa0)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 69 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4CNPCSmBvuG0pERw:-KaFNSTG5d4wkqjh9C4D:b-xsdal6) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L69)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

It's definitely true that `LeafSystem` must be correct for this unit test to pass.  However, isn't there some benefit to protect `LcmPublisherSystem` against future changes in `LeafSystem`, which this unit test perhaps does?
</blockquote></details>

Maybe?  What failure mode would this test find, that the next test wouldn't?  And how likely would LeafSystem be to entertain a regression of that nature?

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF7hvdeFz3W8n3J4RW:-KaFNdhoWk3k3wK0HSFt:b-81nlyc) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L162)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I'll move these unit tests into the other file once they are done being reviewed.
</blockquote></details>

SGTM.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 4 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 86 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF0J8hhx2aS1Z4J4st:-KaFPMjDdbloJUcRDL1I:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L86)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW The "This pointer" vs "The pointer" is now inconsistent in the different constructors and factory.  And FWIW, I think "The" was better, because "This" means `this` (the method receiver) sometimes.  But at least be consistent within the file one way or the other if you are going to rephrase it.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 104 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF1iZECumHx2NLN3ea:-KaFNv--0OXgHwTTXO4e:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L104)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The final sentence is not quite true, I think.  Infinity means "whatever the simulator does by default" or something, not "every major timestep". 

Perhaps consider, rather than rewriting some API language and getting it wrong and/or being brittle, maybe just specify this method by citing the LeafSystem method's contract that it exposes?  It's not like we're transforming or hiding the semantics; we are merely electing to allow callers to govern this, rather than keeping it encapsulated within the System, as e.g., a device driver would want to do.
</blockquote></details>

Done.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 108 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF2-XY_w-28IgccM3K:-KaFNxewZccKazhgOfE3:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L108)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Shouldn't the TODO be placed at the implementation, not the declaration?
</blockquote></details>

Removed.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 113 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF275f7M0sEqCgmKxc:-KaFN5v9AqSCvFSg6CSf:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/lcm_publisher_system.h#L113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Again, probably just worth deleting this, since nobody needs it!  They can write and test it if and when they do.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 69 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4CNPCSmBvuG0pERw:-KaFTVElCEte0eSbfaqP:b-8smcjs) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Maybe?  What failure mode would this test find, that the next test wouldn't?  And how likely would LeafSystem be to entertain a regression of that nature?
</blockquote></details>

One thing the test above covers that the one below does not cover is the ability to switch publish rates at a non-zero sim time. Should I try to fold that into the test below?

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 156 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF6QcQkSB-2WkZkW2Y:-KaFTm3hAzspa_-HV9wL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L156)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

These three lines are repeated 8 times and are all very copy-pasta'y.  Convince me that this phrasing is better than a loop over a data table, or even just checking that `(timestamp * 1e-3) == floor(get_time / kPublishPeriod)` during a for-loop-swept range of StepTo targets?  That would be like 5 lines of code.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 69 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4CNPCSmBvuG0pERw:-KaFU6XT4Gyz2D-t5-zY:bkfcibn) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L69)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

One thing the test above covers that the one below does not cover is the ability to switch publish rates at a non-zero sim time. Should I try to fold that into the test below?
</blockquote></details>

IMO changing on the fly is not worth supporting even during simulation, nevermind as unit testing for a 1-line method.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF7hvdeFz3W8n3J4RW:-KaFVh79C98IKriAPDSH:b-t1kn00) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L162)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

SGTM.
</blockquote></details>

The second case is okay to move now.  I have some more details in terms of unused includes or whatever, but they will be moot / changed once it moves, so I can tackle it in the new location.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 111 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFVM8TaQgmv-ZdlqGu:-KaFVM8UZ099azc04sdk:bscml8) ([raw file](https://github.com/robotlocomotion/drake/blob/a48bec9f80551896b032f3f2271791d2bd1f7863/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L111)):*
> ```cpp
>     EXPECT_EQ(simulator.get_mutable_context()->get_time(), time);
>     const double expected_time =
>         floor(time / kPublishPeriod) * kPublishPeriod * 1000;
> ```

`std::floor`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 69 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF4CNPCSmBvuG0pERw:-KaFdBg-Wxy6mCop76O8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

IMO changing on the fly is not worth supporting even during simulation, nevermind as unit testing for a 1-line method.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 162 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF7hvdeFz3W8n3J4RW:-KaFdDBqPuOUXZsZH74z:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47ef4a0e666d9937b632c0a0e798e6e96287fe2d/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L162)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The second case is okay to move now.  I have some more details in terms of unused includes or whatever, but they will be moot / changed once it moves, so I can tackle it in the new location.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc, line 111 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFVM8TaQgmv-ZdlqGu:-KaFcj6MjHWGM9HqTISh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a48bec9f80551896b032f3f2271791d2bd1f7863/drake/systems/lcm/test/lcm_publisher_system_publish_rate_test.cc#L111)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`std::floor`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaF3HOzMQUe3eMFxJfs:-KaFdaUxfh8YDVBM3ljL:b-zhpi24) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This PR has gone through so many rewrites that I'm sure that I'm blind to some defects by now.  Once the current discussions are resolved, please invite David or Sherm as an extra, fresh reviewer.
</blockquote></details>

OK will do.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 106 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFfTAngyKtExTwpd3g:-KaFfTAngyKtExTwpd3h:b-400uzn) ([raw file](https://github.com/robotlocomotion/drake/blob/8162f653e4cfedfffbb9b55420e46e26aa2958a7/drake/systems/lcm/lcm_publisher_system.h#L106)):*
> ```objc
>    * parameter `period`.
>    */
>   void set_publish_period(double period);
> ```

BTW Wouldn't this be clearer after `make_name`, right next to DoPublish and keeping get_channel_name and make_name adjacent?

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 161 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFflRdzSJhvDvZw963:-KaFflRe1jDVHZQ7XJ4j:b-9ooso6) ([raw file](https://github.com/robotlocomotion/drake/blob/8162f653e4cfedfffbb9b55420e46e26aa2958a7/drake/systems/lcm/test/lcm_publisher_system_test.cc#L161)):*
> ```cpp
>   EXPECT_EQ(transmitted_message.decode(transmitted_message_bytes.data(), 0,
>                            transmitted_message_bytes.size()),
>             transmitted_message_bytes.size());
> ```

FYI For the future perhaps, this decoding logic would be great to have as a template method in `DrakeMockLcm` for everyone to use during testing, so that this method wouldn't be needed and the call site below would just be replaced with `EXPECT_EQ(lcm.DecodeLastPublishedMessageAs<lcmt_drake_signal>().timestamp, expected_time);`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaFgTdsz6tIa-nAY5iN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@david-german-tri for second platform review please due to this PR having numerous major revisions. Thanks!

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 106 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFfTAngyKtExTwpd3g:-KaHLhW91jbql_mPulcV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8162f653e4cfedfffbb9b55420e46e26aa2958a7/drake/systems/lcm/lcm_publisher_system.h#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Wouldn't this be clearer after `make_name`, right next to DoPublish and keeping get_channel_name and make_name adjacent?
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_publisher_system_test.cc, line 161 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFflRdzSJhvDvZw963:-KaHLvIido_KfQqm4sYe:bmfj9w) ([raw file](https://github.com/robotlocomotion/drake/blob/8162f653e4cfedfffbb9b55420e46e26aa2958a7/drake/systems/lcm/test/lcm_publisher_system_test.cc#L161)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI For the future perhaps, this decoding logic would be great to have as a template method in `DrakeMockLcm` for everyone to use during testing, so that this method wouldn't be needed and the call site below would just be replaced with `EXPECT_EQ(lcm.DecodeLastPublishedMessageAs<lcmt_drake_signal>().timestamp, expected_time);`
</blockquote></details>

Agreed. I'll do that as a followup PR: #4760.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaHN53-U_tPjAHFibyZ:b6hs7d2)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 106 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFfTAngyKtExTwpd3g:-KaHczGYQgBUjZNW55__:b-wcz65n) ([raw file](https://github.com/robotlocomotion/drake/blob/8162f653e4cfedfffbb9b55420e46e26aa2958a7/drake/systems/lcm/lcm_publisher_system.h#L106)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

No change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/lcm/lcm_publisher_system.h, line 106 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-KaFfTAngyKtExTwpd3g:-KaHrk6mJSrvogRJLcgt:bprrw02) ([raw file](https://github.com/robotlocomotion/drake/blob/8162f653e4cfedfffbb9b55420e46e26aa2958a7/drake/systems/lcm/lcm_publisher_system.h#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

No change?
</blockquote></details>

Sorry. I forgot to commit / push. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r2, 2 of 4 files at r7, 1 of 1 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4731#-:-KaIME8ckqaSY-qprtGL:bnfp4nl)*
<!-- Sent from Reviewable.io -->
